### PR TITLE
fix: AppStart Transaction for Apps Using UIScenes

### LINKED ISSUE #1375: No startup transaction for ViewController
### Environment

Latest Cocoa SDK 7.4.3

### Steps to Reproduce

Launch iOS-Swift Sample App

### Expected Result

Transaction for ViewController should appear in Sentry.

### Actual Result

No transaction shows up.

### PR DESCRIPTION
**Pull request by brustolin**
## :scroll: Description

Swizzle the rootViewController from the first UIScene to appear in the app in the case UIApplicationDelegate does not have an UIWindow, for iOS 13 and above.

## :bulb: Motivation and Context

In iOS 13 and above, `UIWindow` may be manage by `UIScene`.
We need to Swizzle the first UIViewController from the first UIScene to be created, otherwise we lose transaction for the app main ViewController.

Closes #1375

## :green_heart: How did you test it?

Unit tests

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

**Comment by brustolin** [comment]
@phi

> Awesome, I like the idea. It's similar to the app delegate approach. Can we please add unit tests and I think we should also add UI tests? Is this working when an app has multiple scenes simultaneously from the start?

Hey @philipphofmann, I was thinking about the UI test, and I believe, what we can do with this is open the app to see if the app will not crash with this change. This we already have. Am I missing something?

File: Tests/SentryTests/Integrations/Performance/UIViewController/SentryUIViewControllerSwizzlingTests.swift [hunk 1]
```diff
@@ -66,6 +71,50 @@ class SentryUIViewControllerSwizzlingTests: XCTestCase {
...
+        
+        let notification = Notification(name: NSNotification.Name(rawValue: "UISceneWillConnectNotification"), object: mockWindowScene)
+        swizzler.swizzleRootViewController(fromSceneDelegateNotification: notification)
+        
+        XCTAssertEqual(swizzler.viewControllers.count, 1)
+        XCTAssertTrue(swizzler.viewControllers[0] is TestViewController)
+    }
+    
+    @available(iOS 13.0, *)
+    func testSwizzle_fromScene_invalideNotification() {
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/UIViewController/SentryUIViewControllerSwizzlingTests.swift:91]: `l`: Small typo
```suggestion
    func testSwizzle_fromScene_invalidNotification() {
```

File: Tests/SentryTests/Integrations/Performance/UIViewController/SentryUIViewControllerSwizzlingTests.swift [hunk 2]
```diff
@@ -66,6 +71,50 @@ class SentryUIViewControllerSwizzlingTests: XCTestCase {
...
+        let swizzler = TestSentryUIViewControllerSwizziling(options: fixture.options, dispatchQueue: TestSentryDispatchQueueWrapper())
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/UIViewController/SentryUIViewControllerSwizzlingTests.swift:92]:
```suggestion
        let swizzler = TestSentryUIViewControllerSwizzling(options: fixture.options, dispatchQueue: TestSentryDispatchQueueWrapper())
```

File: Tests/SentryTests/Integrations/Performance/UIViewController/SentryUIViewControllerSwizzlingTests.swift [hunk 3]
```diff
@@ -66,6 +71,50 @@ class SentryUIViewControllerSwizzlingTests: XCTestCase {
...
+        
+        notification = Notification(name: NSNotification.Name(rawValue: "UISceneWillConnectNotification"), object: "Other type of Object")
+        swizzler.swizzleRootViewController(fromSceneDelegateNotification: notification)
+        
+        XCTAssertEqual(swizzler.viewControllers.count, 0)
+        
+        notification = Notification(name: NSNotification.Name(rawValue: "UISceneWillConnectNotification"), object: ObjectWithWindowsProperty(resultOfWindows: "Windows property of the wrong type"))
+        swizzler.swizzleRootViewController(fromSceneDelegateNotification: notification)
+        
+        XCTAssertEqual(swizzler.viewControllers.count, 0)
```
Comment (philipphofmann) [inline Tests/SentryTests/Integrations/Performance/UIViewController/SentryUIViewControllerSwizzlingTests.swift:116]: `m`: Nice test, but I think it would make sense to split this up into several test cases with proper namings as this tests many different things.

**Review by philipphofmann — COMMENTED** [review]
I just realized that `SentryUIViewControllerSwizziling` has quite some not tested code. I can help you with adding more tests. We should have 100% test coverage for the class before merging this PR to make sure not to break something.

**Comment by philipphofmann** [comment]
> @phi
> 
> > Awesome, I like the idea. It's similar to the app delegate approach. Can we please add unit tests and I think we should also add UI tests? Is this working when an app has multiple scenes simultaneously from the start?
> 
> Hey @philipphofmann, I was thinking about the UI test, and I believe, what we can do with this is open the app to see if the app will not crash with this change. This we already have. Am I missing something?

I think adding some samples to the iOS-Swift sample app on extra screens and opening them in a UI test to make sure we don't crash is the minimum we have to do. We can add more tests later on.

File: Sources/Sentry/SentryUIViewControllerSwizziling.m [hunk 4]
```diff
@@ -102,37 +102,112 @@ - (void)swizzleRootViewController
         return;
     }
 
+    if (![self swizzleRootViewControllerFromAppDelegate:app]) {
+        NSString *message
+            = @"UIViewControllerSwizziling: Fail to find root UIViewController from "
+              @"UIApplicationDelegate. Trying to use UISceneWillConnectNotification notification.";
+        [SentryLog logWithMessage:message andLevel:kSentryLevelDebug];
```
[the remark this answers is not in the document]
Comment (brustolin) [reply Sources/Sentry/SentryUIViewControllerSwizziling.m:109]: 😉

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,7 @@
 
 - fix: SentryOptions initWithDict type errors (#1443)
 - fix: Transaction default status should be OK (#1439)
+- fix: Swizzling rootViewController from UIScenes (#1427)
```
Comment (philipphofmann) [inline CHANGELOG.md:7]: Way better 👏

**Review by philipphofmann — COMMENTED** [review]
I just tried it with the iOS-Swift sample app, and I didn't get any app start transactions. Maybe this is not working correctly?

**Comment by brustolin** [comment]
> I just tried it with the iOS-Swift sample app, and I didn't get any app start transactions. Maybe this is not working correctly?

🤔 
Weird.
Let me investigate that.

**Comment by brustolin** [comment]
@philipphofmann I did a wrong assumption and repeat the assumption on tests. Is fixed now.

File: CHANGELOG.md (hunk 5 above)
Comment (bruno-garcia) [reply CHANGELOG.md:7]: We should add a new entry `# Unreleased` and this item under it.
This is changing a released version

**Review by philipphofmann — APPROVED** [review]
LGTM, but we should fix the changelog before merging.
